src/ast/str_util.rs

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+    loop {
+        start += indention(&input[start..]).take(indent).map(char::len_utf8).sum::<usize>();
+        let end = input[start..].find('\n').map(|i| start + i + 1);
+        {
+            let end = end.unwrap_or(input.len());
+            output.push_str(&input[start..end]);
+        }
+        start = match end {
+            Some(end) => end,
+            None => break,
+        };
+    }
+    output
+}
+
+/// Remove any trailing whitespace from a string
+pub fn remove_trailing(string: &mut String) {
+    let trailing: usize = string
+        .chars()
+        .rev()
+        .take_while(|&c| c != '\n' && c.is_whitespace())
+        .map(char::len_utf8)
+        .sum();
+    let len = string.len();
+    string.truncate(len - trailing);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn string_unescapes() {
+        assert_eq!(unescape(r#"Hello\n\"World\" :D"#, false), "Hello\n\"World\" :D");
+        assert_eq!(unescape(r#"Hello''\n'''World''' :D"#, true), "Hello\n''World'' :D");
+    }
+    #[test]
+    fn string_remove_common_indent() {
+        assert_eq!(remove_common_indent("\n  \n    \n \n "), "\n\n\n");
+        assert_eq!(remove_common_indent("\n  \n    \n a\n"), " \n   \na\n");
+        assert_eq!(remove_common_indent("  \n    \n a\n"), "   \na\n");
+    }
+    #[test]
+    fn parts() {
+        use crate::{NixLanguage, SyntaxNode};
+        use rowan::{GreenNodeBuilder, Language};
+
+        fn string_node(content: &str) -> ast::Str {
+            let mut builder = GreenNodeBuilder::new();
+            builder.start_node(NixLanguage::kind_to_raw(NODE_STRING));
+            builder.token(NixLanguage::kind_to_raw(TOKEN_STRING_START), "''");
+            builder.token(NixLanguage::kind_to_raw(TOKEN_STRING_CONTENT), content);
+            builder.token(NixLanguage::kind_to_raw(TOKEN_STRING_END), "''");
+            builder.finish_node();
+
+            ast::Str::cast(SyntaxNode::new_root(builder.finish())).unwrap()
+        }
+
+        let txtin = r#"
+                        |trailing-whitespace
+                              |trailing-whitespace
+                    This is a multiline string :D
+                      indented by two
+                    \'\'\'\'\
+                    ''${ interpolation was escaped }
+                    two single quotes: '''
+                    three single quotes: ''''
+                "#
+        .replace("|trailing-whitespace", "");
+
+        if let [StrPart::Literal(lit)] = &ast::Str::string_parts(&string_node(txtin.as_str()))[..] {
+            assert_eq!(lit,
+                // Get the below with nix repl
+                "    \n          \nThis is a multiline string :D\n  indented by two\n\\'\\'\\'\\'\\\n${ interpolation was escaped }\ntwo single quotes: ''\nthree single quotes: '''\n"
+            );
+        } else {
+            unreachable!();
+        }
+    }
+
+    //     #[test]
+    //     fn parts_ast() {
```

Comment:
This test checked for unintentional truncation (which happens when the string unescaping parser breaks out too early). This intend should be covered by a test to prevent it from re-occuring (see also #66)